### PR TITLE
Introduce PHP7.2 `object` type hint

### PR DESCRIPTION
**Pull request by shulard**
Hello,

This PR is about introducing the PHP7.2 `object` type hint recently merged in the PHP core as requested in #110.

### CODE REVIEW

File: src/Generator/TypeGenerator.php [hunk 1]
```diff
@@ -33,7 +33,8 @@
      *
      * @link http://php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration
      */
-    private static $internalPhpTypes = ['void', 'int', 'float', 'string', 'bool', 'array', 'callable', 'iterable'];
+    private static $internalPhpTypes = ['void', 'int', 'float', 'string', 'bool', 'array', 'callable', 'iterable',
+        'object'];
```
Comment (Ocramius) [inline src/Generator/TypeGenerator.php]: Alignment is broken here

File: test/Generator/MethodGeneratorTest.php [hunk 2]
```diff
@@ -373,7 +376,7 @@ function (array $parameter) {
                 return PHP_VERSION_ID >= 70100
                     || (
                         false === strpos($parameter[2], '?')
-                        && ! in_array(strtolower($parameter[2]), ['void', 'iterable'])
+                        && ! in_array(strtolower($parameter[2]), ['void', 'iterable', 'object'])
```
Comment (Ocramius) [inline test/Generator/MethodGeneratorTest.php]: Careful about the php version comparison: should be 70200

File: test/Generator/ParameterGeneratorTest.php [hunk 3]
```diff
@@ -472,13 +473,19 @@ public function reflectionHintsProvider()
...
             [IterableHintsClass::class, 'nullDefaultIterableParameter', 'foo', '?iterable'],
+            [ObjectHintsClass::class, 'objectParameter', 'foo', 'object'],
+            [ObjectHintsClass::class, 'nullableObjectParameter', 'foo', '?object'],
+            [ObjectHintsClass::class, 'nullDefaultObjectParameter', 'foo', '?object'],
         ];
 
         $compatibleParameters = array_filter(
             $parameters,
             function (array $parameter) {
                 return PHP_VERSION_ID >= 70100
```
Comment (Ocramius) [inline test/Generator/ParameterGeneratorTest.php]: Same as above: different conditionals for 70200

File: test/Generator/ParameterGeneratorTest.php [hunk 4]
```diff
@@ -472,13 +473,19 @@ public function reflectionHintsProvider()
             [IterableHintsClass::class, 'iterableParameter', 'foo', 'iterable'],
             [IterableHintsClass::class, 'nullableIterableParameter', 'foo', '?iterable'],
             [IterableHintsClass::class, 'nullDefaultIterableParameter', 'foo', '?iterable'],
+            [ObjectHintsClass::class, 'objectParameter', 'foo', 'object'],
```
Comment (Ocramius) [inline test/Generator/ParameterGeneratorTest.php]: These should only be added to the data providrr if the php version is 70200 or greater

File: test/Generator/MethodGeneratorTest.php [hunk 5]
```diff
@@ -365,6 +366,8 @@ public function returnTypeHintClassesProvider()
             [NullableReturnTypeHintedClass::class, 'otherClassReturn', '?\\' . InternalHintsClass::class],
             [IterableHintsClass::class, 'iterableReturnValue', 'iterable'],
             [IterableHintsClass::class, 'nullableIterableReturnValue', '?iterable'],
+            [ObjectHintsClass::class, 'objectReturnValue', 'object'],
```
Comment (Ocramius) [inline test/Generator/MethodGeneratorTest.php]: These should only be added to the data providrr if the php version is 70200 or greater

File: src/Generator/TypeGenerator.php (hunk 1 above)
Comment (shulard) [inline src/Generator/TypeGenerator.php]: What's the right alignment here ? I try to find a "too" long array property but can't find one in the project...

Is this ok?

```
private static $internalPhpTypes = [
    'void', 
    'int', 
    'float', 
    'string', 
    'bool', 
    'array', 
    'callable', 
    'iterable',
    'object'
];
```

**Comment by Ocramius** [comment]
Yup

File: test/Generator/MethodGeneratorTest.php (hunk 5 above)
Comment (shulard) [inline test/Generator/MethodGeneratorTest.php]: I've updated the tests regarding PHP version.

**Comment by Ocramius** [comment]
Couple things here: 

1. Patch target should be changed to `develop`
2. #114 was just merged, and that may cause rebase conflicts

**Comment by shulard** [comment]
Hello,

I've updated the target branch and rebased on the `develop` branch 😄.

File: test/Generator/MethodGeneratorTest.php [hunk 6]
```diff
@@ -364,9 +365,24 @@ public function returnTypeHintClassesProvider()
...
 
-        return $parameters;
+        return array_filter(
+            $parameters,
+            function (array $parameter) {
+                return (PHP_VERSION_ID >= 70200
+                    || (
+                        false === strpos($parameter[2], 'object')
+                    ))
+                    && (PHP_VERSION_ID >= 70100
```
Comment (webimpress) [inline test/Generator/MethodGeneratorTest.php]: We don't need this part of condition, as now `develop` supports only PHP 7.1+

File: test/Generator/ParameterGeneratorTest.php [hunk 7]
```diff
@@ -464,8 +462,27 @@ public function reflectionHintsProvider()
...
 
+        $compatibleParameters = array_filter(
+            $parameters,
+            function (array $parameter) {
+
+                return (PHP_VERSION_ID >= 70200
+                    || (
+                        false === strpos($parameter[3], 'object')
+                    ))
+                    && (PHP_VERSION_ID >= 70100
```
Comment (webimpress) [inline test/Generator/ParameterGeneratorTest.php]: As above.

**Comment by webimpress** [comment]
@Ocramius tests still are failing !

**Comment by Ocramius** [comment]
Hmm, can you check up on the test failures?

**Comment by shulard** [comment]
I've just removed the PHP 7.1 conditions in the tests...

File: test/Generator/ParameterGeneratorTest.php [hunk 8]
```diff
@@ -464,8 +462,22 @@ public function reflectionHintsProvider()
             [IterableHintsClass::class, 'iterableParameter', 'foo', 'iterable'],
             [IterableHintsClass::class, 'nullableIterableParameter', 'foo', '?iterable'],
             [IterableHintsClass::class, 'nullDefaultIterableParameter', 'foo', '?iterable'],
+            [ObjectHintsClass::class, 'objectParameter', 'foo', 'object'],
+            [ObjectHintsClass::class, 'nullableObjectParameter', 'foo', '?object'],
+            [ObjectHintsClass::class, 'nullDefaultObjectParameter', 'foo', '?object'],
         ];
 
+        $compatibleParameters = array_filter(
```
Comment (webimpress) [inline test/Generator/ParameterGeneratorTest.php]: You have to use this variable in line 487 and 489.

File: test/Generator/ParameterGeneratorTest.php [hunk 9]
```diff
@@ -464,8 +462,22 @@ public function reflectionHintsProvider()
...
+            $parameters,
+            function (array $parameter) {
+
```
Comment (webimpress) [inline test/Generator/ParameterGeneratorTest.php]: Please remove empty line above.

File: test/Generator/TypeGeneratorTest.php [hunk 10]
```diff
@@ -241,6 +241,9 @@ public function invalidTypeProvider()
             ['\\iterable'],
             ['\\Iterable'],
             ['\\ITERABLE'],
+            ['\\object'],
+            ['\\Object'],
+            ['\\OBJECT'],
```
Comment (webimpress) [inline test/Generator/TypeGeneratorTest.php]: Here you added these to `invalidTypeProvider`, don't we need remove 'object' from `validTypeProvider`?

**Comment by shulard** [comment]
Hello @webimpress,

I've updated the code with your comments and fixed the test suite. I've made a mistake in the "validTypeProvider" method...

**Comment by Ocramius** [comment]
:ship:

**Pull request merged**